File: Algorithmen/3x+1/fortran/3x+1.f90
program three_x_plus_one
  implicit none

  !boilerplate
  integer(kind=8) :: n, count, i
  integer :: start_time, elapsed_time
  integer :: percent
  
  !prompt user for input
  write(*, *) "Enter a number:"
  read(*, *) n

  i = n
  count = 0
  call system_clock(start_time)

  !calculate
  do while (i /= 1)
    count = count + 1
    if (mod(i, 2) == 0) then
      i = i / 2
    else
      i = 3 * i + 1
    end if
  end do

  !print result
  call system_clock(elapsed_time)
  elapsed_time = (elapsed_time - start_time) * 1000
  write(*, *) "Starting number: ", n
  write(*, *) "Number of iterations: ", count
  write(*, *) "Execution time: ", elapsed_time, "ms"

end program three_x_plus_one

File: Algorithmen/3x+1/go/3x+1.go
package main

import (
	"fmt"
	"time"
)

func main() {
	var n int
	fmt.Println("Enter a number:")
	fmt.Scan(&n)

	max_n := 0
	max_count := 0
	start_time := time.Now().UnixNano()

	count := 0
	var i int64 = int64(n)
	for i > 1 {
		if i%2 == 0 {
			i = i / 2
		} else {
			i = 3*i + 1
		}
		count++
	}
	max_n = n
	max_count = count

	end_time := time.Now().UnixNano()

	fmt.Printf("Starting number: %d\n", n)
	fmt.Printf("Number of iterations: %d\n", max_count)
	fmt.Printf("Execution time: %d ms\n", (end_time-start_time)/1000000)
}
